src/Orchard.Web/Modules/Orchard.OpenId/Module.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using Orchard.Environment.Shell;
+using Orchard.Security;
+using Orchard.OpenId.Indexes;
+using Orchard.OpenId.Models;
+using Orchard.Users.Models;
+using Orchard.OpenId.Services;
+using Orchard.FileSystem.AppData;
+using Microsoft.Extensions.Logging;
+using Orchard.Settings;
+
+namespace Orchard.OpenId
+{
+    public class Startup : StartupBase
+    {     
+        private readonly string _certificateFullPath;
+        private const string certificateFileName = "Certificate.pfx";
+        private readonly string tenant;
+
+        public Startup(ShellSettings shellSettings, IAppDataFolder appDataFolder, ILoggerFactory loggerFactory)   
+        {
+            tenant = shellSettings.Name;
+            _certificateFullPath = appDataFolder.Combine(shellSettings.Name, certificateFileName);
+        }
+
+        public override void Configure(IApplicationBuilder builder, IRouteBuilder routes, IServiceProvider serviceProvider)
+        {
+            var baseUrl = serviceProvider.GetRequiredService<ISiteService>().GetSiteSettingsAsync().Result.BaseUrl ?? tenant;
+            builder.UseOpenIddict();
```

Comment:
@sebastienros Well I took a look to Startup.cs in Orchard.Demo module class but I didn't see anything different than what is in Module.cs in Orchard.OpenId module that also derives from IStartup.
I asked ISiteService though Startup constructor but it fails, however using serviceProvider.GetRequiredService() it works but by default a tenant doesn't have a base Url. 
I though Configure and ConfigureService is called for each tenant having the opportunity of configuring middleware for each tenant. Is this what you mean? If I'm right then using baseUrl won't be a problem because each tenant will have different Authority and Audience when I use the baseUrl instead of the hardcoded url "http://localhost:2918".
I thought current request is not available at the moment of middleware configuration. I will try. But it will be same request for all tenant middlewares inicializataion, isn't it? So, as you point using same url for all the tenants doesn't seem a good idea.
Am I missing something?